dali/operators/video/reader/video_reader_decoder_op.cc

```diff
@@ -0,0 +1,555 @@
+// Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include "dali/core/boundary.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/span.h"
+
+#include "dali/operators/reader/reader_op.h"
+#include "dali/operators/video/frames_decoder_base.h"
+#include "dali/operators/video/frames_decoder_cpu.h"
+#include "dali/operators/video/frames_decoder_gpu.h"
+#include "dali/operators/video/video_utils.h"
+
+#include "libavutil/rational.h"
+
+namespace dali {
+
+struct VideoSampleDesc {
+  VideoSampleDesc(std::string filename = {}, int label = -1, int start = -1, int end = -1,
+                  int stride = -1)
+      : filename_(filename), label_(label), start_(start), end_(end), stride_(stride) {}
+  std::string filename_;
+  int label_ = -1;
+  int start_ = -1;
+  int end_ = -1;
+  int stride_ = -1;
+};
+
+template <typename Backend>
+struct VideoSample : public VideoSampleDesc {
+  VideoSample(std::string filename = {}, int label = -1, int start = -1, int end = -1,
+              int stride = -1)
+      : VideoSampleDesc{filename, label, start, end, stride} {}
+
+  VideoSample(const VideoSampleDesc &other) noexcept
+      : VideoSampleDesc(other) {}
+
+  // to be filled by Prefetch
+  Tensor<Backend> data_;
+  std::vector<double> timestamps_;
+  std::vector<int64_t> frame_idx_;
+};
+
+enum class FileListFormat {
+  kFrameIndex,
+  kTimestamp,
+  kTimestampInclusive
+};
+
+template <typename Backend, typename FramesDecoderImpl, typename Sample = VideoSample<Backend>>
+class VideoLoaderDecoder : public Loader<Backend, Sample, true> {
+ public:
+  explicit inline VideoLoaderDecoder(const OpSpec &spec)
+      : Loader<Backend, Sample, true>(spec),
+        file_root_(spec.GetArgument<std::string>("file_root")),
+        file_list_(spec.GetArgument<std::string>("file_list")),
+        filenames_(spec.GetRepeatedArgument<std::string>("filenames")),
+        sequence_len_(spec.GetArgument<int>("sequence_length")),
+        stride_(spec.GetArgument<int>("stride")),
+        step_(spec.GetArgument<int>("step")),
+        boundary_type_(GetBoundaryType(spec)) {
+    if ((spec.HasArgument("file_list") || spec.HasArgument("file_root") || spec.HasArgument("filenames")) != 1) {
+      DALI_FAIL("Only one of the following arguments can be provided: ``file_list``, ``file_root``, ``filenames``");
```

Comment:
I don't get this part. The `if` checks if any of the `file*` arguments was provided. If none was, the error message says that "only one can be provided". Looks as if the intention was to check if more than one was provided.

(Nitpick: the error message in the current form might suggest that no other arguments are allowed)

Reply:
it should be `+`. It should be checking that only one of those is provided.